lib/internal/errors.js

```diff
@@ -0,0 +1,139 @@
+'use strict';
+
+// The whole point behind this internal module is to allow Node.js to no
+// longer be forced to treat every error message change as a semver-major
+// change. The NodeError classes here all expose a `code` property whose
+// value statically and permanently identifies the error. While the error
+// message may change, the code should not.
+
+const kCode = Symbol('code');
+const kSymbolsMap = new Map();
+
+function symbolFor(name) {
+  var sym = kSymbolsMap.get(name);
+  if (sym === undefined) {
+    sym = Symbol(name);
+    kSymbolsMap.set(name, sym);
+  }
+  return sym;
+}
+
+var assert, util;
+function lazyAssert() {
+  if (!assert)
+    assert = require('assert');
+  return assert;
+}
+
+function lazyUtil() {
+  if (!util)
+    util = require('util');
+  return util;
+}
+
+class NodeError extends Error {
+  constructor(key /**, ...args **/) {
+    lazyAssert().strictEqual(typeof key, 'string');
+    const args = Array(arguments.length - 1);
```

Comment:
It's possible that things have improved with the most recent v8 drops. I'm running the `benchmark/es/restparams-bench.js` right now to compare